src/gadgets/tower_extension/fq6.rs

```diff
@@ -0,0 +1,345 @@
+use std::sync::Arc;
+
+use pairing::ff::PrimeField;
+
+use super::{fq2::Fq2, params::Extension6Params};
+
+use crate::{
+    cs::traits::cs::ConstraintSystem,
+    field::SmallField,
+    gadgets::{boolean::Boolean, non_native_field::traits::NonNativeField},
+};
+
+/// `Fq6` field extension implementation in the constraint system. It is implemented
+/// as `Fq2[v]/(v^3-xi)` where `xi=9+u`. In other words,
+/// it is a set of quadratic polynomials of a form `c0+c1*v+c2*v^2`,
+///  where `c0`, `c1`, `c2` are elements of `Fq2`.
+/// See https://hackmd.io/@jpw/bn254#Field-extension-towers for reference. For
+/// implementation reference, see https://eprint.iacr.org/2006/471.pdf.
+#[derive(Clone, Debug, Copy)]
+pub struct Fq6<F, T, NN, P>
+where
+    F: SmallField,
+    T: PrimeField,
+    NN: NonNativeField<F, T>,
+    P: Extension6Params<T>,
+{
+    pub c0: Fq2<F, T, NN, P::Ex2>,
+    pub c1: Fq2<F, T, NN, P::Ex2>,
+    pub c2: Fq2<F, T, NN, P::Ex2>,
+    _marker: std::marker::PhantomData<(F, T)>,
+}
+
+impl<F, T, NN, P> Fq6<F, T, NN, P>
+where
+    F: SmallField,
+    T: pairing::ff::PrimeField,
+    NN: NonNativeField<F, T>,
+    P: Extension6Params<T>,
+{
+    /// Creates a new `Fq6` element from three `Fq2` components.
+    pub fn new(
+        c0: Fq2<F, T, NN, P::Ex2>,
+        c1: Fq2<F, T, NN, P::Ex2>,
+        c2: Fq2<F, T, NN, P::Ex2>,
+    ) -> Self {
+        Self {
+            c0,
+            c1,
+            c2,
+            _marker: std::marker::PhantomData::<(F, T)>,
+        }
+    }
+
+    /// Creates a new zero `Fq6` in a form `0+0*v+0*v^2`
+    pub fn zero<CS>(cs: &mut CS, params: &Arc<NN::Params>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let zero = Fq2::zero(cs, params);
+        Self::new(zero.clone(), zero.clone(), zero)
+    }
+
+    /// Creates a unit `Fq6` in a form `1+0*v+0*v^2`
+    pub fn one<CS>(cs: &mut CS, params: &Arc<NN::Params>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let one = Fq2::one(cs, params);
+        let zero = Fq2::zero(cs, params);
+        Self::new(one, zero.clone(), zero)
+    }
+
+    /// Returns true if the `Fq6` element is zero.
+    pub fn is_zero<CS>(&mut self, cs: &mut CS) -> Boolean<F>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let is_c0_zero = self.c0.is_zero(cs);
+        let is_c1_zero = self.c1.is_zero(cs);
+        let is_c2_zero = self.c2.is_zero(cs);
+        is_c0_zero.and(cs, is_c1_zero).and(cs, is_c2_zero)
```

Comment:
`Boolean::multi_and` may work better here, I think its marginally cheaper

Reply:
Yeah, besides, this portion looks much more natural now!